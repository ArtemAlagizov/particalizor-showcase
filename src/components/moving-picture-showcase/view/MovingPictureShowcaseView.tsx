import React from "react";
import { IMovingPictureShowcaseView } from "./IMovingPictureShowcaseView";
import { isNil } from "ramda";
import Dropzone from "react-dropzone";
import { MovingPicture } from "particalizor-3000";
import { Styled } from "../styled";
import "antd/dist/antd.css";
import cadillac from "../../../assets/testImage.png";
import SliderWithLabel from "../../../ui-components/slider-with-label/container/SliderWithLabel";
import DropdownWithLabel from "../../../ui-components/dropdown-with-label/container/DropdownWithLabel";
import SwitchWithLabel from "../../../ui-components/switch-with-label/container/SwitchWithLabel";

const MovingPictureShowcaseView: React.FC<IMovingPictureShowcaseView> = ({
  onParticleNumberChange,
  onTraceWidthChange,
  onVelocityChange,
  onLifeTimeChange,
  onDirectionChannelChange,
  onHueChannelChange,
  onHueReverseChange,
  onDirectionReverseChange,

  onDrop,

  particleNumberValue,
  particleTraceWidthValue,
  particleVelocityValue,
  particleLifeTimeValue,
  directionChannelValue,
  hueChannelValue,
  hueReversedValue,
  directionReversedValue,

  directionHueOptions,
  switchImageMode,
  switchButtonTitle,
  selectedImageMode,
  uploadedImage
}) => {
  return (
    <Styled.ShowcaseComponentContainer>
      <Styled.PropertiesContainer>
        <Styled.PropertiesSet>
          <SliderWithLabel
            propertyName={"particleNumber"}
            selectedValue={particleNumberValue}
            onValueChange={onParticleNumberChange}
            label={"particle number"}
            range={{ min: 200, max: 7000 }}
            step={200}
          />
          <SliderWithLabel
            propertyName={"particleTraceWidthValue"}
            selectedValue={particleTraceWidthValue}
            onValueChange={onTraceWidthChange}
            label={"particle trace width"}
            range={{ min: 1, max: 10 }}
          />
          <SliderWithLabel
            propertyName={"particleVelocityValue"}
            selectedValue={particleVelocityValue}
            onValueChange={onVelocityChange}
            label={"particle velocity"}
            range={{ min: 1, max: 10 }}
          />
          <SliderWithLabel
            propertyName={"particleLifeTimeValue"}
            selectedValue={particleLifeTimeValue}
            onValueChange={onLifeTimeChange}
            label={"particle life time"}
            range={{ min: 100, max: 1000 }}
            step={100}
          />
        </Styled.PropertiesSet>
        <Styled.PropertiesSet>
          <DropdownWithLabel
            propertyName={"directionChannel"}
            onValueChange={onDirectionChannelChange}
            selectedValue={directionChannelValue}
            label={"direction"}
            defaultValue={directionChannelValue}
            options={directionHueOptions}
          />
          <DropdownWithLabel
            propertyName={"hueChannel"}
            onValueChange={onHueChannelChange}
            selectedValue={hueChannelValue}
            label={"hue"}
            defaultValue={hueChannelValue}
            options={directionHueOptions}
          />
          <SwitchWithLabel
            propertyName={"hueReverse"}
            onValueChange={onHueReverseChange}
            label={"hue reversed"}
            defaultValue={hueReversedValue}
          />
          <SwitchWithLabel
            propertyName={"directionReverse"}
            onValueChange={onDirectionReverseChange}
            label={"direction reversed"}
            defaultValue={directionReversedValue}
          />
        </Styled.PropertiesSet>
      </Styled.PropertiesContainer>
      <Styled.ImageModeSwitch onClick={switchImageMode}>
        {switchButtonTitle}
      </Styled.ImageModeSwitch>
      {selectedImageMode === "uploaded-picture" && isNil(uploadedImage) && (
        <Styled.Images>
          <Styled.ImageContainer>
            <Styled.Dashed>
              <Dropzone onDrop={onDrop}>
                {({ getRootProps, getInputProps }) => (
                  <section>
                    <Styled.DropZone {...getRootProps()}>
                      <input {...getInputProps()} />
                      <Styled.DropImageText>
                        DRAG AND DROP AN IMAGE HERE, OR CLICK TO SELECT AN IMAGE
                      </Styled.DropImageText>
                    </Styled.DropZone>
                  </section>
                )}
              </Dropzone>
            </Styled.Dashed>
          </Styled.ImageContainer>
        </Styled.Images>
      )}
      {selectedImageMode === "uploaded-picture" && !isNil(uploadedImage) && (
        <Styled.Images>
          <Styled.ImageContainer>
            <Styled.Dashed>
              <Dropzone onDrop={onDrop}>
                {({ getRootProps, getInputProps }) => (
                  <section>
                    <div {...getRootProps()}>
                      <input {...getInputProps()} />
                      <Styled.ImageContainerImg
                        src={URL.createObjectURL(uploadedImage)}
                        alt={"dropped"}
                      />
                    </div>
                  </section>
                )}
              </Dropzone>
            </Styled.Dashed>
          </Styled.ImageContainer>
          <Styled.ImageContainer>
            <MovingPicture
              key={Date.now()}
              imageSource={URL.createObjectURL(uploadedImage)}
              particleTraceWidth={particleTraceWidthValue}
              particleVelocity={particleVelocityValue}
              particleNumber={particleNumberValue}
              particleLifeTime={particleLifeTimeValue}
              directionChannel={directionChannelValue}
              hueChannel={hueChannelValue}
              reverseHue={hueReversedValue}
              reverseDirection={directionReversedValue}
            />
          </Styled.ImageContainer>
        </Styled.Images>
      )}
      {selectedImageMode === "given-picture" && (
        <Styled.Images>
          <Styled.ImageContainer>
            <Styled.ImageContainerImg src={cadillac} alt={"original"} />
          </Styled.ImageContainer>
          <Styled.ImageContainer>
            <MovingPicture
              key={Date.now()}
              imageSource={cadillac}
              particleTraceWidth={particleTraceWidthValue}
              particleVelocity={particleVelocityValue}
              particleNumber={particleNumberValue}
              particleLifeTime={particleLifeTimeValue}
              directionChannel={directionChannelValue}
              hueChannel={hueChannelValue}
              reverseHue={hueReversedValue}
              reverseDirection={directionReversedValue}
            />
          </Styled.ImageContainer>
        </Styled.Images>
      )}
    </Styled.ShowcaseComponentContainer>
  );
};

export default MovingPictureShowcaseView;
