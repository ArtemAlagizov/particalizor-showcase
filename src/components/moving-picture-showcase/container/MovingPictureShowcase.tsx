import React, { useState } from "react";
import MovingPictureShowcaseView from "../view/MovingPictureShowcaseView";
import { SliderValue } from "antd/es/slider";

const MovingPictureShowcase: React.FC = () => {
  const [
    selectedParticleNumberValue,
    setSelectedParticleNumberValue
  ] = useState<number>(2400);
  const [
    selectedParticleTraceWidthValue,
    setSelectedParticleTraceWidthValue
  ] = useState<number>(2);
  const [
    selectedParticleVelocityValue,
    setSelectedParticleVelocityValue
  ] = useState<number>(1);
  const [
    selectedParticleLifeTimeValue,
    setSelectedParticleLifeTimeValue
  ] = useState<number>(100);
  const [
    selectedDirectionChannelValue,
    setSelectedDirectionChannelValue
  ] = useState<string>("saturation");
  const [selectedHueChannelValue, setSelectedHueChannelValue] = useState<
    string
  >("blue");
  const [selectedHueReversedValue, setSelectedHueReversedValue] = useState<
    boolean
  >(false);
  const [
    selectedDirectionReversedValue,
    setSelectedDirectionReversedValue
  ] = useState<boolean>(false);
  const [uploadedImage, setUploadedImage] = useState<File | null>(null);
  const [selectedImageMode, setSelectedImageMode] = useState<string>(
    "given-picture"
  );

  const onChange = (value: SliderValue) =>
    setSelectedParticleNumberValue(value as number);

  const onTraceWidthChange = (value: SliderValue) =>
    setSelectedParticleTraceWidthValue(value as number);

  const onVelocityChange = (value: SliderValue) =>
    setSelectedParticleVelocityValue(value as number);

  const onLifeTimeChange = (value: SliderValue) =>
    setSelectedParticleLifeTimeValue(value as number);

  const onDirectionChannelChange = (value: string) =>
    setSelectedDirectionChannelValue(value);

  const onHueChannelChange = (value: string) => {
    setSelectedHueChannelValue(value);
  };
  const onHueReverseChange = (value: boolean) =>
    setSelectedHueReversedValue(value);

  const onDirectionReverseChange = (value: boolean) =>
    setSelectedDirectionReversedValue(value);

  const directionHueOptions = [
    {
      value: "hue",
      title: "hue"
    },
    {
      value: "saturation",
      title: "saturation"
    },
    {
      value: "light",
      title: "light"
    },
    {
      value: "red",
      title: "red"
    },
    {
      value: "green",
      title: "green"
    },
    {
      value: "blue",
      title: "blue"
    }
  ];

  const switchImageMode = () => {
    setSelectedImageMode(
      selectedImageMode === "given-picture"
        ? "uploaded-picture"
        : "given-picture"
    );
  };

  const switchButtonTitle =
    selectedImageMode === "given-picture"
      ? "THIS IS FIXED PICTURE MODE, CLICK TO SEE UPLOADED PICTURE MODE"
      : "THIS IS UPLOADED PICTURE MODE, CLICK TO SEE FIXED PICTURE MODE";

  const onDrop = (acceptedFiles: File[]) => setUploadedImage(acceptedFiles[0]);

  return (
    <MovingPictureShowcaseView
      directionChannelValue={selectedDirectionChannelValue}
      directionHueOptions={directionHueOptions}
      directionReversedValue={selectedDirectionReversedValue}
      hueChannelValue={selectedHueChannelValue}
      hueReversedValue={selectedHueReversedValue}
      onDirectionChannelChange={onDirectionChannelChange}
      onDirectionReverseChange={onDirectionReverseChange}
      onHueChannelChange={onHueChannelChange}
      onHueReverseChange={onHueReverseChange}
      onLifeTimeChange={onLifeTimeChange}
      onParticleNumberChange={onChange}
      onTraceWidthChange={onTraceWidthChange}
      onVelocityChange={onVelocityChange}
      particleLifeTimeValue={selectedParticleLifeTimeValue}
      particleNumberValue={selectedParticleNumberValue}
      particleTraceWidthValue={selectedParticleTraceWidthValue}
      particleVelocityValue={selectedParticleVelocityValue}
      selectedImageMode={selectedImageMode}
      onDrop={onDrop}
      switchButtonTitle={switchButtonTitle}
      switchImageMode={switchImageMode}
      uploadedImage={uploadedImage}
    />
  );
};

export default MovingPictureShowcase;
