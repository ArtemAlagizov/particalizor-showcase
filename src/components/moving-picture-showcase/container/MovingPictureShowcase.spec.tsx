import React from "react";
import { mount, shallow } from "enzyme";
import MovingPictureShowcase from "./MovingPictureShowcase";
import MovingPictureShowcaseView from "../view/MovingPictureShowcaseView";

describe("Given a MovingPictureShowcase", () => {
  describe("when rendering", () => {
    it("should match snapshot", () => {
      const component = shallow(<MovingPictureShowcase />);

      expect(component).toMatchSnapshot();
    });
    it("should render MovingPictureView", () => {
      const component = mount(<MovingPictureShowcase />);

      expect(component.find(MovingPictureShowcaseView).length).toBe(1);
    });
  });
});
