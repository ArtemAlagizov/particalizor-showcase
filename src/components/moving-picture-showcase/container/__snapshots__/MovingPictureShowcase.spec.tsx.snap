// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Given a MovingPictureShowcase when rendering should match snapshot 1`] = `
<MovingPictureShowcaseView
  directionChannelValue="saturation"
  directionHueOptions={
    Array [
      Object {
        "title": "hue",
        "value": "hue",
      },
      Object {
        "title": "saturation",
        "value": "saturation",
      },
      Object {
        "title": "light",
        "value": "light",
      },
      Object {
        "title": "red",
        "value": "red",
      },
      Object {
        "title": "green",
        "value": "green",
      },
      Object {
        "title": "blue",
        "value": "blue",
      },
    ]
  }
  directionReversedValue={false}
  hueChannelValue="blue"
  hueReversedValue={false}
  onDirectionChannelChange={[Function]}
  onDirectionReverseChange={[Function]}
  onDrop={[Function]}
  onHueChannelChange={[Function]}
  onHueReverseChange={[Function]}
  onLifeTimeChange={[Function]}
  onParticleNumberChange={[Function]}
  onTraceWidthChange={[Function]}
  onVelocityChange={[Function]}
  particleLifeTimeValue={100}
  particleNumberValue={2400}
  particleTraceWidthValue={2}
  particleVelocityValue={1}
  selectedImageMode="given-picture"
  switchButtonTitle="THIS IS FIXED PICTURE MODE, CLICK TO SEE UPLOADED PICTURE MODE"
  switchImageMode={[Function]}
  uploadedImage={null}
/>
`;
